{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Porting DESeq into python using rpy2#\n",
    "\n",
    "I will use a small example of [ERCC transcript](https://www.thermofisher.com/order/catalog/product/4456740) from [samples A and B in MAQC data](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3272078/)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "import pandas as pd \n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will read the table and it should only contains count data of ERCC spikeins (rows) and 3 replicates from each of samples A and B (columns)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>A_1</th>\n",
       "      <th>A_2</th>\n",
       "      <th>A_3</th>\n",
       "      <th>B_1</th>\n",
       "      <th>B_2</th>\n",
       "      <th>B_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERCC-00002</td>\n",
       "      <td>111461</td>\n",
       "      <td>106261</td>\n",
       "      <td>107547</td>\n",
       "      <td>333944</td>\n",
       "      <td>199252</td>\n",
       "      <td>186947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERCC-00003</td>\n",
       "      <td>6735</td>\n",
       "      <td>5387</td>\n",
       "      <td>5265</td>\n",
       "      <td>13937</td>\n",
       "      <td>8584</td>\n",
       "      <td>8596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERCC-00004</td>\n",
       "      <td>17673</td>\n",
       "      <td>13983</td>\n",
       "      <td>15462</td>\n",
       "      <td>5065</td>\n",
       "      <td>3222</td>\n",
       "      <td>3353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERCC-00009</td>\n",
       "      <td>4669</td>\n",
       "      <td>4431</td>\n",
       "      <td>4211</td>\n",
       "      <td>6939</td>\n",
       "      <td>4155</td>\n",
       "      <td>3647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERCC-00012</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           id     A_1     A_2     A_3     B_1     B_2     B_3\n",
       "0  ERCC-00002  111461  106261  107547  333944  199252  186947\n",
       "1  ERCC-00003    6735    5387    5265   13937    8584    8596\n",
       "2  ERCC-00004   17673   13983   15462    5065    3222    3353\n",
       "3  ERCC-00009    4669    4431    4211    6939    4155    3647\n",
       "4  ERCC-00012       0       2       0       0       0       0"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_table('../test/data/ercc.tsv')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And here, we will create a design matrix based on the samples in the count table. Note that the sample name has to be used as the ```pd.DataFrame``` index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>samplename</th>\n",
       "      <th>sample</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>samplename</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A_1</th>\n",
       "      <td>A_1</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_2</th>\n",
       "      <td>A_2</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_3</th>\n",
       "      <td>A_3</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B_1</th>\n",
       "      <td>B_1</td>\n",
       "      <td>B</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B_2</th>\n",
       "      <td>B_2</td>\n",
       "      <td>B</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B_3</th>\n",
       "      <td>B_3</td>\n",
       "      <td>B</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           samplename sample replicate\n",
       "samplename                            \n",
       "A_1               A_1      A         1\n",
       "A_2               A_2      A         2\n",
       "A_3               A_3      A         3\n",
       "B_1               B_1      B         1\n",
       "B_2               B_2      B         2\n",
       "B_3               B_3      B         3"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_df = pd.DataFrame({'samplename': df.columns}) \\\n",
    "        .query('samplename != \"id\"')\\\n",
    "        .assign(sample = lambda d: d.samplename.str.extract('([AB])_', expand=False)) \\\n",
    "        .assign(replicate = lambda d: d.samplename.str.extract('_([123])', expand=False)) \n",
    "sample_df.index = sample_df.samplename\n",
    "sample_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Running DESeq2 is jsut like how it is run in ```R```, but instead of the row.name being gene ID for the count table, we can jsut tell the function which column is the gene ID:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating size factors\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating dispersions\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: mean-dispersion relationship\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: final dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: fitting model and testing\n",
      "\n",
      "INFO:DESeq2:Using contrast: ['sample', 'B', 'A']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERCC-00002</th>\n",
       "      <td>167917.342729</td>\n",
       "      <td>0.808857</td>\n",
       "      <td>0.047606</td>\n",
       "      <td>16.990537</td>\n",
       "      <td>9.650176e-65</td>\n",
       "      <td>1.102877e-63</td>\n",
       "      <td>ERCC-00002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERCC-00003</th>\n",
       "      <td>7902.634073</td>\n",
       "      <td>0.521731</td>\n",
       "      <td>0.058878</td>\n",
       "      <td>8.861252</td>\n",
       "      <td>7.912104e-19</td>\n",
       "      <td>4.868987e-18</td>\n",
       "      <td>ERCC-00003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERCC-00004</th>\n",
       "      <td>10567.048228</td>\n",
       "      <td>-2.330122</td>\n",
       "      <td>0.055754</td>\n",
       "      <td>-41.792764</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>ERCC-00004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERCC-00009</th>\n",
       "      <td>4672.573043</td>\n",
       "      <td>-0.195660</td>\n",
       "      <td>0.061600</td>\n",
       "      <td>-3.176286</td>\n",
       "      <td>1.491736e-03</td>\n",
       "      <td>3.616329e-03</td>\n",
       "      <td>ERCC-00009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERCC-00012</th>\n",
       "      <td>0.384257</td>\n",
       "      <td>-1.565491</td>\n",
       "      <td>4.047562</td>\n",
       "      <td>-0.386774</td>\n",
       "      <td>6.989237e-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERCC-00012</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 baseMean  log2FoldChange     lfcSE       stat        pvalue  \\\n",
       "ERCC-00002  167917.342729        0.808857  0.047606  16.990537  9.650176e-65   \n",
       "ERCC-00003    7902.634073        0.521731  0.058878   8.861252  7.912104e-19   \n",
       "ERCC-00004   10567.048228       -2.330122  0.055754 -41.792764  0.000000e+00   \n",
       "ERCC-00009    4672.573043       -0.195660  0.061600  -3.176286  1.491736e-03   \n",
       "ERCC-00012       0.384257       -1.565491  4.047562  -0.386774  6.989237e-01   \n",
       "\n",
       "                    padj          id  \n",
       "ERCC-00002  1.102877e-63  ERCC-00002  \n",
       "ERCC-00003  4.868987e-18  ERCC-00003  \n",
       "ERCC-00004  0.000000e+00  ERCC-00004  \n",
       "ERCC-00009  3.616329e-03  ERCC-00009  \n",
       "ERCC-00012           NaN  ERCC-00012  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from diffexpr.py_deseq import py_DESeq2\n",
    "\n",
    "dds = py_DESeq2(count_matrix = df,\n",
    "               design_matrix = sample_df,\n",
    "               design_formula = '~ replicate + sample',\n",
    "               gene_column = 'id') # <- telling DESeq2 this should be the gene ID column\n",
    "    \n",
    "dds.run_deseq() \n",
    "dds.get_deseq_result(contrast = ['sample','B','A'])\n",
    "res = dds.deseq_result \n",
    "res.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:DESeq2:Normalizing counts\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A_1</th>\n",
       "      <th>A_2</th>\n",
       "      <th>A_3</th>\n",
       "      <th>B_1</th>\n",
       "      <th>B_2</th>\n",
       "      <th>B_3</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERCC-00002</th>\n",
       "      <td>115018.353297</td>\n",
       "      <td>122494.471246</td>\n",
       "      <td>128809.545168</td>\n",
       "      <td>218857.357008</td>\n",
       "      <td>207880.854689</td>\n",
       "      <td>214443.474968</td>\n",
       "      <td>ERCC-00002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERCC-00003</th>\n",
       "      <td>6949.952086</td>\n",
       "      <td>6209.970889</td>\n",
       "      <td>6305.915138</td>\n",
       "      <td>9133.911628</td>\n",
       "      <td>8955.740754</td>\n",
       "      <td>9860.313944</td>\n",
       "      <td>ERCC-00003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERCC-00004</th>\n",
       "      <td>18237.045763</td>\n",
       "      <td>16119.180051</td>\n",
       "      <td>18518.909755</td>\n",
       "      <td>3319.456296</td>\n",
       "      <td>3361.532701</td>\n",
       "      <td>3846.164804</td>\n",
       "      <td>ERCC-00004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERCC-00009</th>\n",
       "      <td>4818.014297</td>\n",
       "      <td>5107.922964</td>\n",
       "      <td>5043.534405</td>\n",
       "      <td>4547.622357</td>\n",
       "      <td>4334.937422</td>\n",
       "      <td>4183.406812</td>\n",
       "      <td>ERCC-00009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERCC-00012</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.305540</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ERCC-00012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERCC-00164</th>\n",
       "      <td>2.063831</td>\n",
       "      <td>1.152770</td>\n",
       "      <td>5.988523</td>\n",
       "      <td>3.276857</td>\n",
       "      <td>1.043306</td>\n",
       "      <td>2.294163</td>\n",
       "      <td>ERCC-00164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERCC-00165</th>\n",
       "      <td>269.329992</td>\n",
       "      <td>246.692736</td>\n",
       "      <td>287.449123</td>\n",
       "      <td>513.811202</td>\n",
       "      <td>484.094095</td>\n",
       "      <td>489.803869</td>\n",
       "      <td>ERCC-00165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERCC-00168</th>\n",
       "      <td>1.031916</td>\n",
       "      <td>3.458309</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.587600</td>\n",
       "      <td>4.173225</td>\n",
       "      <td>1.147082</td>\n",
       "      <td>ERCC-00168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERCC-00170</th>\n",
       "      <td>137.244785</td>\n",
       "      <td>148.707304</td>\n",
       "      <td>135.340629</td>\n",
       "      <td>26.870229</td>\n",
       "      <td>10.433062</td>\n",
       "      <td>32.118286</td>\n",
       "      <td>ERCC-00170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERCC-00171</th>\n",
       "      <td>8707.304484</td>\n",
       "      <td>9622.169484</td>\n",
       "      <td>8818.699555</td>\n",
       "      <td>7691.439109</td>\n",
       "      <td>7691.253592</td>\n",
       "      <td>6892.813691</td>\n",
       "      <td>ERCC-00171</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      A_1            A_2            A_3            B_1  \\\n",
       "ERCC-00002  115018.353297  122494.471246  128809.545168  218857.357008   \n",
       "ERCC-00003    6949.952086    6209.970889    6305.915138    9133.911628   \n",
       "ERCC-00004   18237.045763   16119.180051   18518.909755    3319.456296   \n",
       "ERCC-00009    4818.014297    5107.922964    5043.534405    4547.622357   \n",
       "ERCC-00012       0.000000       2.305540       0.000000       0.000000   \n",
       "...                   ...            ...            ...            ...   \n",
       "ERCC-00164       2.063831       1.152770       5.988523       3.276857   \n",
       "ERCC-00165     269.329992     246.692736     287.449123     513.811202   \n",
       "ERCC-00168       1.031916       3.458309       0.000000       4.587600   \n",
       "ERCC-00170     137.244785     148.707304     135.340629      26.870229   \n",
       "ERCC-00171    8707.304484    9622.169484    8818.699555    7691.439109   \n",
       "\n",
       "                      B_2            B_3          id  \n",
       "ERCC-00002  207880.854689  214443.474968  ERCC-00002  \n",
       "ERCC-00003    8955.740754    9860.313944  ERCC-00003  \n",
       "ERCC-00004    3361.532701    3846.164804  ERCC-00004  \n",
       "ERCC-00009    4334.937422    4183.406812  ERCC-00009  \n",
       "ERCC-00012       0.000000       0.000000  ERCC-00012  \n",
       "...                   ...            ...         ...  \n",
       "ERCC-00164       1.043306       2.294163  ERCC-00164  \n",
       "ERCC-00165     484.094095     489.803869  ERCC-00165  \n",
       "ERCC-00168       4.173225       1.147082  ERCC-00168  \n",
       "ERCC-00170      10.433062      32.118286  ERCC-00170  \n",
       "ERCC-00171    7691.253592    6892.813691  ERCC-00171  \n",
       "\n",
       "[92 rows x 7 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dds.normalized_count() #DESeq2 normalized count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Intercept', 'replicate_2_vs_1', 'replicate_3_vs_1', 'sample_B_vs_A']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dds.comparison # show coefficients for GLM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERCC-00002</td>\n",
       "      <td>167917.342729</td>\n",
       "      <td>0.807316</td>\n",
       "      <td>0.047609</td>\n",
       "      <td>9.650176e-65</td>\n",
       "      <td>1.102877e-63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERCC-00003</td>\n",
       "      <td>7902.634073</td>\n",
       "      <td>0.519944</td>\n",
       "      <td>0.058823</td>\n",
       "      <td>7.912104e-19</td>\n",
       "      <td>4.868987e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERCC-00004</td>\n",
       "      <td>10567.048228</td>\n",
       "      <td>-2.328037</td>\n",
       "      <td>0.055783</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERCC-00009</td>\n",
       "      <td>4672.573043</td>\n",
       "      <td>-0.194594</td>\n",
       "      <td>0.061466</td>\n",
       "      <td>1.491736e-03</td>\n",
       "      <td>3.616329e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERCC-00012</td>\n",
       "      <td>0.384257</td>\n",
       "      <td>-0.052326</td>\n",
       "      <td>0.820696</td>\n",
       "      <td>6.989237e-01</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           id       baseMean  log2FoldChange     lfcSE        pvalue  \\\n",
       "0  ERCC-00002  167917.342729        0.807316  0.047609  9.650176e-65   \n",
       "1  ERCC-00003    7902.634073        0.519944  0.058823  7.912104e-19   \n",
       "2  ERCC-00004   10567.048228       -2.328037  0.055783  0.000000e+00   \n",
       "3  ERCC-00009    4672.573043       -0.194594  0.061466  1.491736e-03   \n",
       "4  ERCC-00012       0.384257       -0.052326  0.820696  6.989237e-01   \n",
       "\n",
       "           padj  \n",
       "0  1.102877e-63  \n",
       "1  4.868987e-18  \n",
       "2  0.000000e+00  \n",
       "3  3.616329e-03  \n",
       "4           NaN  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# from the last cell, we see the arrangement of coefficients, \n",
    "# so that we can now use \"coef\" for lfcShrink\n",
    "# the comparison we want to focus on is 'sample_B_vs_A', so coef = 4 will be used\n",
    "lfc_res = dds.lfcShrink(coef=4, method='apeglm')\n",
    "lfc_res.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
