{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Porting DESeq into python using rpy2#\n",
    "\n",
    "I will use a small example of [ERCC transcript](https://www.thermofisher.com/order/catalog/product/4456740) from [samples A and B in MAQC data](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3272078/)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "import pandas as pd \n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will read the table and it should only contains count data of ERCC spikeins (rows) and 3 replicates from each of samples A and B (columns)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TargetName</th>\n",
       "      <th>disease3_scan | 001 | PanCK</th>\n",
       "      <th>disease3_scan | 001 | neg</th>\n",
       "      <th>disease3_scan | 002 | PanCK</th>\n",
       "      <th>disease3_scan | 002 | neg</th>\n",
       "      <th>disease3_scan | 003 | PanCK</th>\n",
       "      <th>disease3_scan | 003 | neg</th>\n",
       "      <th>disease3_scan | 004 | PanCK</th>\n",
       "      <th>disease3_scan | 004 | neg</th>\n",
       "      <th>disease3_scan | 005 | PanCK</th>\n",
       "      <th>...</th>\n",
       "      <th>disease1B_scan | 015 | Geometric Segment</th>\n",
       "      <th>disease1B_scan | 016 | Geometric Segment</th>\n",
       "      <th>disease1B_scan | 017 | Geometric Segment</th>\n",
       "      <th>disease1B_scan | 018 | Geometric Segment</th>\n",
       "      <th>disease1B_scan | 019 | Geometric Segment</th>\n",
       "      <th>disease1B_scan | 020 | Geometric Segment</th>\n",
       "      <th>disease1B_scan | 021 | Geometric Segment</th>\n",
       "      <th>disease1B_scan | 022 | Geometric Segment</th>\n",
       "      <th>disease1B_scan | 023 | Geometric Segment</th>\n",
       "      <th>disease1B_scan | 024 | Geometric Segment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PADI2</td>\n",
       "      <td>15.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CYP24A1</td>\n",
       "      <td>23.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SUPT16H</td>\n",
       "      <td>35.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZMIZ2</td>\n",
       "      <td>51.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>87.0</td>\n",
       "      <td>113.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>84.0</td>\n",
       "      <td>164.0</td>\n",
       "      <td>118.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SPAG9</td>\n",
       "      <td>48.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>91.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>...</td>\n",
       "      <td>8.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 232 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  TargetName  disease3_scan | 001 | PanCK  disease3_scan | 001 | neg  \\\n",
       "0      PADI2                         15.0                       35.0   \n",
       "1    CYP24A1                         23.0                       32.0   \n",
       "2    SUPT16H                         35.0                       66.0   \n",
       "3      ZMIZ2                         51.0                       81.0   \n",
       "4      SPAG9                         48.0                       65.0   \n",
       "\n",
       "   disease3_scan | 002 | PanCK  disease3_scan | 002 | neg  \\\n",
       "0                         36.0                       41.0   \n",
       "1                         23.0                       41.0   \n",
       "2                         46.0                       61.0   \n",
       "3                         87.0                      113.0   \n",
       "4                         46.0                       91.0   \n",
       "\n",
       "   disease3_scan | 003 | PanCK  disease3_scan | 003 | neg  \\\n",
       "0                          9.0                       37.0   \n",
       "1                          3.0                       30.0   \n",
       "2                         20.0                       39.0   \n",
       "3                         68.0                       84.0   \n",
       "4                         25.0                       58.0   \n",
       "\n",
       "   disease3_scan | 004 | PanCK  disease3_scan | 004 | neg  \\\n",
       "0                         42.0                       37.0   \n",
       "1                         34.0                       32.0   \n",
       "2                         79.0                       66.0   \n",
       "3                        164.0                      118.0   \n",
       "4                         95.0                       81.0   \n",
       "\n",
       "   disease3_scan | 005 | PanCK  ...  disease1B_scan | 015 | Geometric Segment  \\\n",
       "0                         14.0  ...                                       1.0   \n",
       "1                         17.0  ...                                       1.0   \n",
       "2                         43.0  ...                                       1.0   \n",
       "3                         81.0  ...                                       3.0   \n",
       "4                         45.0  ...                                       8.0   \n",
       "\n",
       "   disease1B_scan | 016 | Geometric Segment  \\\n",
       "0                                       1.0   \n",
       "1                                       1.0   \n",
       "2                                       3.0   \n",
       "3                                       2.0   \n",
       "4                                      10.0   \n",
       "\n",
       "   disease1B_scan | 017 | Geometric Segment  \\\n",
       "0                                       3.0   \n",
       "1                                       1.0   \n",
       "2                                       2.0   \n",
       "3                                       2.0   \n",
       "4                                       7.0   \n",
       "\n",
       "   disease1B_scan | 018 | Geometric Segment  \\\n",
       "0                                       1.0   \n",
       "1                                       1.0   \n",
       "2                                       1.0   \n",
       "3                                       1.0   \n",
       "4                                       3.0   \n",
       "\n",
       "   disease1B_scan | 019 | Geometric Segment  \\\n",
       "0                                       1.0   \n",
       "1                                       1.0   \n",
       "2                                       1.0   \n",
       "3                                       1.0   \n",
       "4                                       1.0   \n",
       "\n",
       "   disease1B_scan | 020 | Geometric Segment  \\\n",
       "0                                       1.0   \n",
       "1                                       1.0   \n",
       "2                                       3.0   \n",
       "3                                       3.0   \n",
       "4                                       7.0   \n",
       "\n",
       "   disease1B_scan | 021 | Geometric Segment  \\\n",
       "0                                       1.0   \n",
       "1                                       3.0   \n",
       "2                                       1.0   \n",
       "3                                       2.0   \n",
       "4                                       5.0   \n",
       "\n",
       "   disease1B_scan | 022 | Geometric Segment  \\\n",
       "0                                       2.0   \n",
       "1                                       1.0   \n",
       "2                                       2.0   \n",
       "3                                       2.0   \n",
       "4                                       4.0   \n",
       "\n",
       "   disease1B_scan | 023 | Geometric Segment  \\\n",
       "0                                       1.0   \n",
       "1                                       3.0   \n",
       "2                                       1.0   \n",
       "3                                       3.0   \n",
       "4                                       6.0   \n",
       "\n",
       "   disease1B_scan | 024 | Geometric Segment  \n",
       "0                                       1.0  \n",
       "1                                       1.0  \n",
       "2                                       1.0  \n",
       "3                                       2.0  \n",
       "4                                       2.0  \n",
       "\n",
       "[5 rows x 232 columns]"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_table('../test/data/ercc1.txt')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TargetName</th>\n",
       "      <th>disease3_scan | 001 | PanCK</th>\n",
       "      <th>disease3_scan | 001 | neg</th>\n",
       "      <th>disease3_scan | 002 | PanCK</th>\n",
       "      <th>disease3_scan | 002 | neg</th>\n",
       "      <th>disease3_scan | 003 | PanCK</th>\n",
       "      <th>disease3_scan | 003 | neg</th>\n",
       "      <th>disease3_scan | 004 | PanCK</th>\n",
       "      <th>disease3_scan | 004 | neg</th>\n",
       "      <th>disease3_scan | 005 | PanCK</th>\n",
       "      <th>...</th>\n",
       "      <th>disease1B_scan | 015 | Geometric Segment</th>\n",
       "      <th>disease1B_scan | 016 | Geometric Segment</th>\n",
       "      <th>disease1B_scan | 017 | Geometric Segment</th>\n",
       "      <th>disease1B_scan | 018 | Geometric Segment</th>\n",
       "      <th>disease1B_scan | 019 | Geometric Segment</th>\n",
       "      <th>disease1B_scan | 020 | Geometric Segment</th>\n",
       "      <th>disease1B_scan | 021 | Geometric Segment</th>\n",
       "      <th>disease1B_scan | 022 | Geometric Segment</th>\n",
       "      <th>disease1B_scan | 023 | Geometric Segment</th>\n",
       "      <th>disease1B_scan | 024 | Geometric Segment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PADI2</td>\n",
       "      <td>15</td>\n",
       "      <td>35</td>\n",
       "      <td>36</td>\n",
       "      <td>41</td>\n",
       "      <td>9</td>\n",
       "      <td>37</td>\n",
       "      <td>42</td>\n",
       "      <td>37</td>\n",
       "      <td>14</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CYP24A1</td>\n",
       "      <td>23</td>\n",
       "      <td>32</td>\n",
       "      <td>23</td>\n",
       "      <td>41</td>\n",
       "      <td>3</td>\n",
       "      <td>30</td>\n",
       "      <td>34</td>\n",
       "      <td>32</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SUPT16H</td>\n",
       "      <td>35</td>\n",
       "      <td>66</td>\n",
       "      <td>46</td>\n",
       "      <td>61</td>\n",
       "      <td>20</td>\n",
       "      <td>39</td>\n",
       "      <td>79</td>\n",
       "      <td>66</td>\n",
       "      <td>43</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZMIZ2</td>\n",
       "      <td>51</td>\n",
       "      <td>81</td>\n",
       "      <td>87</td>\n",
       "      <td>113</td>\n",
       "      <td>68</td>\n",
       "      <td>84</td>\n",
       "      <td>164</td>\n",
       "      <td>118</td>\n",
       "      <td>81</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SPAG9</td>\n",
       "      <td>48</td>\n",
       "      <td>65</td>\n",
       "      <td>46</td>\n",
       "      <td>91</td>\n",
       "      <td>25</td>\n",
       "      <td>58</td>\n",
       "      <td>95</td>\n",
       "      <td>81</td>\n",
       "      <td>45</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 232 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  TargetName  disease3_scan | 001 | PanCK  disease3_scan | 001 | neg  \\\n",
       "0      PADI2                           15                         35   \n",
       "1    CYP24A1                           23                         32   \n",
       "2    SUPT16H                           35                         66   \n",
       "3      ZMIZ2                           51                         81   \n",
       "4      SPAG9                           48                         65   \n",
       "\n",
       "   disease3_scan | 002 | PanCK  disease3_scan | 002 | neg  \\\n",
       "0                           36                         41   \n",
       "1                           23                         41   \n",
       "2                           46                         61   \n",
       "3                           87                        113   \n",
       "4                           46                         91   \n",
       "\n",
       "   disease3_scan | 003 | PanCK  disease3_scan | 003 | neg  \\\n",
       "0                            9                         37   \n",
       "1                            3                         30   \n",
       "2                           20                         39   \n",
       "3                           68                         84   \n",
       "4                           25                         58   \n",
       "\n",
       "   disease3_scan | 004 | PanCK  disease3_scan | 004 | neg  \\\n",
       "0                           42                         37   \n",
       "1                           34                         32   \n",
       "2                           79                         66   \n",
       "3                          164                        118   \n",
       "4                           95                         81   \n",
       "\n",
       "   disease3_scan | 005 | PanCK  ...  disease1B_scan | 015 | Geometric Segment  \\\n",
       "0                           14  ...                                         1   \n",
       "1                           17  ...                                         1   \n",
       "2                           43  ...                                         1   \n",
       "3                           81  ...                                         3   \n",
       "4                           45  ...                                         8   \n",
       "\n",
       "   disease1B_scan | 016 | Geometric Segment  \\\n",
       "0                                         1   \n",
       "1                                         1   \n",
       "2                                         3   \n",
       "3                                         2   \n",
       "4                                        10   \n",
       "\n",
       "   disease1B_scan | 017 | Geometric Segment  \\\n",
       "0                                         3   \n",
       "1                                         1   \n",
       "2                                         2   \n",
       "3                                         2   \n",
       "4                                         7   \n",
       "\n",
       "   disease1B_scan | 018 | Geometric Segment  \\\n",
       "0                                         1   \n",
       "1                                         1   \n",
       "2                                         1   \n",
       "3                                         1   \n",
       "4                                         3   \n",
       "\n",
       "   disease1B_scan | 019 | Geometric Segment  \\\n",
       "0                                         1   \n",
       "1                                         1   \n",
       "2                                         1   \n",
       "3                                         1   \n",
       "4                                         1   \n",
       "\n",
       "   disease1B_scan | 020 | Geometric Segment  \\\n",
       "0                                         1   \n",
       "1                                         1   \n",
       "2                                         3   \n",
       "3                                         3   \n",
       "4                                         7   \n",
       "\n",
       "   disease1B_scan | 021 | Geometric Segment  \\\n",
       "0                                         1   \n",
       "1                                         3   \n",
       "2                                         1   \n",
       "3                                         2   \n",
       "4                                         5   \n",
       "\n",
       "   disease1B_scan | 022 | Geometric Segment  \\\n",
       "0                                         2   \n",
       "1                                         1   \n",
       "2                                         2   \n",
       "3                                         2   \n",
       "4                                         4   \n",
       "\n",
       "   disease1B_scan | 023 | Geometric Segment  \\\n",
       "0                                         1   \n",
       "1                                         3   \n",
       "2                                         1   \n",
       "3                                         3   \n",
       "4                                         6   \n",
       "\n",
       "   disease1B_scan | 024 | Geometric Segment  \n",
       "0                                         1  \n",
       "1                                         1  \n",
       "2                                         1  \n",
       "3                                         2  \n",
       "4                                         2  \n",
       "\n",
       "[5 rows x 232 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# df[[\"a\", \"b\"]] = df[[\"a\", \"b\"]].apply(pd.to_numeric)\n",
    "df.loc[:, df.columns != 'TargetName'] = df.loc[:, df.columns != 'TargetName'].astype(int)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And here, we will create a design matrix based on the samples in the count table. Note that the sample name has to be used as the ```pd.DataFrame``` index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SlideName</th>\n",
       "      <th>ScanName</th>\n",
       "      <th>ROILabel</th>\n",
       "      <th>SegmentLabel</th>\n",
       "      <th>SegmentDisplayName</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>AOISurfaceArea</th>\n",
       "      <th>AOINucleiCount</th>\n",
       "      <th>ROICoordinateX</th>\n",
       "      <th>ROICoordinateY</th>\n",
       "      <th>...</th>\n",
       "      <th>SequencingSaturation</th>\n",
       "      <th>UMIQ30</th>\n",
       "      <th>RTSQ30</th>\n",
       "      <th>disease_status</th>\n",
       "      <th>pathology</th>\n",
       "      <th>region</th>\n",
       "      <th>LOQ</th>\n",
       "      <th>NormalizationFactor</th>\n",
       "      <th>RoiReportX</th>\n",
       "      <th>RoiReportY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>disease3</td>\n",
       "      <td>disease3_scan</td>\n",
       "      <td>7</td>\n",
       "      <td>Geometric Segment</td>\n",
       "      <td>disease3_scan | 007 | Geometric Segment</td>\n",
       "      <td>DSP-1001250007851-H-A02</td>\n",
       "      <td>31797.59253</td>\n",
       "      <td>202</td>\n",
       "      <td>11444</td>\n",
       "      <td>-18819</td>\n",
       "      <td>...</td>\n",
       "      <td>90.586744</td>\n",
       "      <td>0.9939</td>\n",
       "      <td>0.9932</td>\n",
       "      <td>DKD</td>\n",
       "      <td>abnormal</td>\n",
       "      <td>glomerulus</td>\n",
       "      <td>24.246524</td>\n",
       "      <td>0.556033</td>\n",
       "      <td>1197</td>\n",
       "      <td>-3538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>disease3</td>\n",
       "      <td>disease3_scan</td>\n",
       "      <td>8</td>\n",
       "      <td>Geometric Segment</td>\n",
       "      <td>disease3_scan | 008 | Geometric Segment</td>\n",
       "      <td>DSP-1001250007851-H-A03</td>\n",
       "      <td>16920.10267</td>\n",
       "      <td>102</td>\n",
       "      <td>13064</td>\n",
       "      <td>-20021</td>\n",
       "      <td>...</td>\n",
       "      <td>90.707085</td>\n",
       "      <td>0.9948</td>\n",
       "      <td>0.9942</td>\n",
       "      <td>DKD</td>\n",
       "      <td>abnormal</td>\n",
       "      <td>glomerulus</td>\n",
       "      <td>19.338643</td>\n",
       "      <td>0.820810</td>\n",
       "      <td>1606</td>\n",
       "      <td>-3834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>disease3</td>\n",
       "      <td>disease3_scan</td>\n",
       "      <td>9</td>\n",
       "      <td>Geometric Segment</td>\n",
       "      <td>disease3_scan | 009 | Geometric Segment</td>\n",
       "      <td>DSP-1001250007851-H-A04</td>\n",
       "      <td>14312.32987</td>\n",
       "      <td>98</td>\n",
       "      <td>14221</td>\n",
       "      <td>-18591</td>\n",
       "      <td>...</td>\n",
       "      <td>87.861873</td>\n",
       "      <td>0.9949</td>\n",
       "      <td>0.9943</td>\n",
       "      <td>DKD</td>\n",
       "      <td>abnormal</td>\n",
       "      <td>glomerulus</td>\n",
       "      <td>17.017533</td>\n",
       "      <td>0.957612</td>\n",
       "      <td>1892</td>\n",
       "      <td>-3474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>disease3</td>\n",
       "      <td>disease3_scan</td>\n",
       "      <td>10</td>\n",
       "      <td>Geometric Segment</td>\n",
       "      <td>disease3_scan | 010 | Geometric Segment</td>\n",
       "      <td>DSP-1001250007851-H-A05</td>\n",
       "      <td>20032.83995</td>\n",
       "      <td>143</td>\n",
       "      <td>12120</td>\n",
       "      <td>-17683</td>\n",
       "      <td>...</td>\n",
       "      <td>89.399676</td>\n",
       "      <td>0.9925</td>\n",
       "      <td>0.9909</td>\n",
       "      <td>DKD</td>\n",
       "      <td>abnormal</td>\n",
       "      <td>glomerulus</td>\n",
       "      <td>22.335012</td>\n",
       "      <td>0.718209</td>\n",
       "      <td>1366</td>\n",
       "      <td>-3245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>disease3</td>\n",
       "      <td>disease3_scan</td>\n",
       "      <td>11</td>\n",
       "      <td>Geometric Segment</td>\n",
       "      <td>disease3_scan | 011 | Geometric Segment</td>\n",
       "      <td>DSP-1001250007851-H-A06</td>\n",
       "      <td>27583.26127</td>\n",
       "      <td>195</td>\n",
       "      <td>12816</td>\n",
       "      <td>-16378</td>\n",
       "      <td>...</td>\n",
       "      <td>89.958230</td>\n",
       "      <td>0.9934</td>\n",
       "      <td>0.9923</td>\n",
       "      <td>DKD</td>\n",
       "      <td>abnormal</td>\n",
       "      <td>glomerulus</td>\n",
       "      <td>24.703838</td>\n",
       "      <td>0.492486</td>\n",
       "      <td>1542</td>\n",
       "      <td>-2925</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  SlideName       ScanName  ROILabel       SegmentLabel  \\\n",
       "0  disease3  disease3_scan         7  Geometric Segment   \n",
       "1  disease3  disease3_scan         8  Geometric Segment   \n",
       "2  disease3  disease3_scan         9  Geometric Segment   \n",
       "3  disease3  disease3_scan        10  Geometric Segment   \n",
       "4  disease3  disease3_scan        11  Geometric Segment   \n",
       "\n",
       "                        SegmentDisplayName                Sample_ID  \\\n",
       "0  disease3_scan | 007 | Geometric Segment  DSP-1001250007851-H-A02   \n",
       "1  disease3_scan | 008 | Geometric Segment  DSP-1001250007851-H-A03   \n",
       "2  disease3_scan | 009 | Geometric Segment  DSP-1001250007851-H-A04   \n",
       "3  disease3_scan | 010 | Geometric Segment  DSP-1001250007851-H-A05   \n",
       "4  disease3_scan | 011 | Geometric Segment  DSP-1001250007851-H-A06   \n",
       "\n",
       "   AOISurfaceArea  AOINucleiCount  ROICoordinateX  ROICoordinateY  ...  \\\n",
       "0     31797.59253             202           11444          -18819  ...   \n",
       "1     16920.10267             102           13064          -20021  ...   \n",
       "2     14312.32987              98           14221          -18591  ...   \n",
       "3     20032.83995             143           12120          -17683  ...   \n",
       "4     27583.26127             195           12816          -16378  ...   \n",
       "\n",
       "   SequencingSaturation  UMIQ30  RTSQ30  disease_status  pathology  \\\n",
       "0             90.586744  0.9939  0.9932             DKD   abnormal   \n",
       "1             90.707085  0.9948  0.9942             DKD   abnormal   \n",
       "2             87.861873  0.9949  0.9943             DKD   abnormal   \n",
       "3             89.399676  0.9925  0.9909             DKD   abnormal   \n",
       "4             89.958230  0.9934  0.9923             DKD   abnormal   \n",
       "\n",
       "       region        LOQ  NormalizationFactor RoiReportX RoiReportY  \n",
       "0  glomerulus  24.246524             0.556033       1197      -3538  \n",
       "1  glomerulus  19.338643             0.820810       1606      -3834  \n",
       "2  glomerulus  17.017533             0.957612       1892      -3474  \n",
       "3  glomerulus  22.335012             0.718209       1366      -3245  \n",
       "4  glomerulus  24.703838             0.492486       1542      -2925  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_df = pd.read_table('../test/data/ercc2.txt')\n",
    "sample_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SlideName</th>\n",
       "      <th>ScanName</th>\n",
       "      <th>ROILabel</th>\n",
       "      <th>SegmentLabel</th>\n",
       "      <th>SegmentDisplayName</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>AOISurfaceArea</th>\n",
       "      <th>AOINucleiCount</th>\n",
       "      <th>ROICoordinateX</th>\n",
       "      <th>ROICoordinateY</th>\n",
       "      <th>...</th>\n",
       "      <th>SequencingSaturation</th>\n",
       "      <th>UMIQ30</th>\n",
       "      <th>RTSQ30</th>\n",
       "      <th>disease_status</th>\n",
       "      <th>pathology</th>\n",
       "      <th>region</th>\n",
       "      <th>LOQ</th>\n",
       "      <th>NormalizationFactor</th>\n",
       "      <th>RoiReportX</th>\n",
       "      <th>RoiReportY</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SegmentDisplayName</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>disease3_scan | 007 | Geometric Segment</th>\n",
       "      <td>disease3</td>\n",
       "      <td>disease3_scan</td>\n",
       "      <td>7</td>\n",
       "      <td>Geometric Segment</td>\n",
       "      <td>disease3_scan | 007 | Geometric Segment</td>\n",
       "      <td>DSP-1001250007851-H-A02</td>\n",
       "      <td>31797.59253</td>\n",
       "      <td>202</td>\n",
       "      <td>11444</td>\n",
       "      <td>-18819</td>\n",
       "      <td>...</td>\n",
       "      <td>90.586744</td>\n",
       "      <td>0.9939</td>\n",
       "      <td>0.9932</td>\n",
       "      <td>DKD</td>\n",
       "      <td>abnormal</td>\n",
       "      <td>glomerulus</td>\n",
       "      <td>24.246524</td>\n",
       "      <td>0.556033</td>\n",
       "      <td>1197</td>\n",
       "      <td>-3538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>disease3_scan | 008 | Geometric Segment</th>\n",
       "      <td>disease3</td>\n",
       "      <td>disease3_scan</td>\n",
       "      <td>8</td>\n",
       "      <td>Geometric Segment</td>\n",
       "      <td>disease3_scan | 008 | Geometric Segment</td>\n",
       "      <td>DSP-1001250007851-H-A03</td>\n",
       "      <td>16920.10267</td>\n",
       "      <td>102</td>\n",
       "      <td>13064</td>\n",
       "      <td>-20021</td>\n",
       "      <td>...</td>\n",
       "      <td>90.707085</td>\n",
       "      <td>0.9948</td>\n",
       "      <td>0.9942</td>\n",
       "      <td>DKD</td>\n",
       "      <td>abnormal</td>\n",
       "      <td>glomerulus</td>\n",
       "      <td>19.338643</td>\n",
       "      <td>0.820810</td>\n",
       "      <td>1606</td>\n",
       "      <td>-3834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>disease3_scan | 009 | Geometric Segment</th>\n",
       "      <td>disease3</td>\n",
       "      <td>disease3_scan</td>\n",
       "      <td>9</td>\n",
       "      <td>Geometric Segment</td>\n",
       "      <td>disease3_scan | 009 | Geometric Segment</td>\n",
       "      <td>DSP-1001250007851-H-A04</td>\n",
       "      <td>14312.32987</td>\n",
       "      <td>98</td>\n",
       "      <td>14221</td>\n",
       "      <td>-18591</td>\n",
       "      <td>...</td>\n",
       "      <td>87.861873</td>\n",
       "      <td>0.9949</td>\n",
       "      <td>0.9943</td>\n",
       "      <td>DKD</td>\n",
       "      <td>abnormal</td>\n",
       "      <td>glomerulus</td>\n",
       "      <td>17.017533</td>\n",
       "      <td>0.957612</td>\n",
       "      <td>1892</td>\n",
       "      <td>-3474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>disease3_scan | 010 | Geometric Segment</th>\n",
       "      <td>disease3</td>\n",
       "      <td>disease3_scan</td>\n",
       "      <td>10</td>\n",
       "      <td>Geometric Segment</td>\n",
       "      <td>disease3_scan | 010 | Geometric Segment</td>\n",
       "      <td>DSP-1001250007851-H-A05</td>\n",
       "      <td>20032.83995</td>\n",
       "      <td>143</td>\n",
       "      <td>12120</td>\n",
       "      <td>-17683</td>\n",
       "      <td>...</td>\n",
       "      <td>89.399676</td>\n",
       "      <td>0.9925</td>\n",
       "      <td>0.9909</td>\n",
       "      <td>DKD</td>\n",
       "      <td>abnormal</td>\n",
       "      <td>glomerulus</td>\n",
       "      <td>22.335012</td>\n",
       "      <td>0.718209</td>\n",
       "      <td>1366</td>\n",
       "      <td>-3245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>disease3_scan | 011 | Geometric Segment</th>\n",
       "      <td>disease3</td>\n",
       "      <td>disease3_scan</td>\n",
       "      <td>11</td>\n",
       "      <td>Geometric Segment</td>\n",
       "      <td>disease3_scan | 011 | Geometric Segment</td>\n",
       "      <td>DSP-1001250007851-H-A06</td>\n",
       "      <td>27583.26127</td>\n",
       "      <td>195</td>\n",
       "      <td>12816</td>\n",
       "      <td>-16378</td>\n",
       "      <td>...</td>\n",
       "      <td>89.958230</td>\n",
       "      <td>0.9934</td>\n",
       "      <td>0.9923</td>\n",
       "      <td>DKD</td>\n",
       "      <td>abnormal</td>\n",
       "      <td>glomerulus</td>\n",
       "      <td>24.703838</td>\n",
       "      <td>0.492486</td>\n",
       "      <td>1542</td>\n",
       "      <td>-2925</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        SlideName       ScanName  ROILabel  \\\n",
       "SegmentDisplayName                                                           \n",
       "disease3_scan | 007 | Geometric Segment  disease3  disease3_scan         7   \n",
       "disease3_scan | 008 | Geometric Segment  disease3  disease3_scan         8   \n",
       "disease3_scan | 009 | Geometric Segment  disease3  disease3_scan         9   \n",
       "disease3_scan | 010 | Geometric Segment  disease3  disease3_scan        10   \n",
       "disease3_scan | 011 | Geometric Segment  disease3  disease3_scan        11   \n",
       "\n",
       "                                              SegmentLabel  \\\n",
       "SegmentDisplayName                                           \n",
       "disease3_scan | 007 | Geometric Segment  Geometric Segment   \n",
       "disease3_scan | 008 | Geometric Segment  Geometric Segment   \n",
       "disease3_scan | 009 | Geometric Segment  Geometric Segment   \n",
       "disease3_scan | 010 | Geometric Segment  Geometric Segment   \n",
       "disease3_scan | 011 | Geometric Segment  Geometric Segment   \n",
       "\n",
       "                                                              SegmentDisplayName  \\\n",
       "SegmentDisplayName                                                                 \n",
       "disease3_scan | 007 | Geometric Segment  disease3_scan | 007 | Geometric Segment   \n",
       "disease3_scan | 008 | Geometric Segment  disease3_scan | 008 | Geometric Segment   \n",
       "disease3_scan | 009 | Geometric Segment  disease3_scan | 009 | Geometric Segment   \n",
       "disease3_scan | 010 | Geometric Segment  disease3_scan | 010 | Geometric Segment   \n",
       "disease3_scan | 011 | Geometric Segment  disease3_scan | 011 | Geometric Segment   \n",
       "\n",
       "                                                       Sample_ID  \\\n",
       "SegmentDisplayName                                                 \n",
       "disease3_scan | 007 | Geometric Segment  DSP-1001250007851-H-A02   \n",
       "disease3_scan | 008 | Geometric Segment  DSP-1001250007851-H-A03   \n",
       "disease3_scan | 009 | Geometric Segment  DSP-1001250007851-H-A04   \n",
       "disease3_scan | 010 | Geometric Segment  DSP-1001250007851-H-A05   \n",
       "disease3_scan | 011 | Geometric Segment  DSP-1001250007851-H-A06   \n",
       "\n",
       "                                         AOISurfaceArea  AOINucleiCount  \\\n",
       "SegmentDisplayName                                                        \n",
       "disease3_scan | 007 | Geometric Segment     31797.59253             202   \n",
       "disease3_scan | 008 | Geometric Segment     16920.10267             102   \n",
       "disease3_scan | 009 | Geometric Segment     14312.32987              98   \n",
       "disease3_scan | 010 | Geometric Segment     20032.83995             143   \n",
       "disease3_scan | 011 | Geometric Segment     27583.26127             195   \n",
       "\n",
       "                                         ROICoordinateX  ROICoordinateY  ...  \\\n",
       "SegmentDisplayName                                                       ...   \n",
       "disease3_scan | 007 | Geometric Segment           11444          -18819  ...   \n",
       "disease3_scan | 008 | Geometric Segment           13064          -20021  ...   \n",
       "disease3_scan | 009 | Geometric Segment           14221          -18591  ...   \n",
       "disease3_scan | 010 | Geometric Segment           12120          -17683  ...   \n",
       "disease3_scan | 011 | Geometric Segment           12816          -16378  ...   \n",
       "\n",
       "                                         SequencingSaturation  UMIQ30  RTSQ30  \\\n",
       "SegmentDisplayName                                                              \n",
       "disease3_scan | 007 | Geometric Segment             90.586744  0.9939  0.9932   \n",
       "disease3_scan | 008 | Geometric Segment             90.707085  0.9948  0.9942   \n",
       "disease3_scan | 009 | Geometric Segment             87.861873  0.9949  0.9943   \n",
       "disease3_scan | 010 | Geometric Segment             89.399676  0.9925  0.9909   \n",
       "disease3_scan | 011 | Geometric Segment             89.958230  0.9934  0.9923   \n",
       "\n",
       "                                         disease_status  pathology  \\\n",
       "SegmentDisplayName                                                   \n",
       "disease3_scan | 007 | Geometric Segment             DKD   abnormal   \n",
       "disease3_scan | 008 | Geometric Segment             DKD   abnormal   \n",
       "disease3_scan | 009 | Geometric Segment             DKD   abnormal   \n",
       "disease3_scan | 010 | Geometric Segment             DKD   abnormal   \n",
       "disease3_scan | 011 | Geometric Segment             DKD   abnormal   \n",
       "\n",
       "                                             region        LOQ  \\\n",
       "SegmentDisplayName                                               \n",
       "disease3_scan | 007 | Geometric Segment  glomerulus  24.246524   \n",
       "disease3_scan | 008 | Geometric Segment  glomerulus  19.338643   \n",
       "disease3_scan | 009 | Geometric Segment  glomerulus  17.017533   \n",
       "disease3_scan | 010 | Geometric Segment  glomerulus  22.335012   \n",
       "disease3_scan | 011 | Geometric Segment  glomerulus  24.703838   \n",
       "\n",
       "                                         NormalizationFactor RoiReportX  \\\n",
       "SegmentDisplayName                                                        \n",
       "disease3_scan | 007 | Geometric Segment             0.556033       1197   \n",
       "disease3_scan | 008 | Geometric Segment             0.820810       1606   \n",
       "disease3_scan | 009 | Geometric Segment             0.957612       1892   \n",
       "disease3_scan | 010 | Geometric Segment             0.718209       1366   \n",
       "disease3_scan | 011 | Geometric Segment             0.492486       1542   \n",
       "\n",
       "                                        RoiReportY  \n",
       "SegmentDisplayName                                  \n",
       "disease3_scan | 007 | Geometric Segment      -3538  \n",
       "disease3_scan | 008 | Geometric Segment      -3834  \n",
       "disease3_scan | 009 | Geometric Segment      -3474  \n",
       "disease3_scan | 010 | Geometric Segment      -3245  \n",
       "disease3_scan | 011 | Geometric Segment      -2925  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_df.index = sample_df.SegmentDisplayName\n",
    "sample_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "231 231\n"
     ]
    }
   ],
   "source": [
    "l2 = set(sample_df.SegmentDisplayName.unique())\n",
    "l1 = set(df.columns[1:].tolist())\n",
    "print(len(l1), len(l2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The lists are identical\n"
     ]
    }
   ],
   "source": [
    "# using == to check if \n",
    "# lists are equal\n",
    "if l1 == l2:\n",
    "    print (\"The lists are identical\")\n",
    "else :\n",
    "    print (\"The lists are not identical\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Running DESeq2 is jsut like how it is run in ```R```, but instead of the row.name being gene ID for the count table, we can jsut tell the function which column is the gene ID:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Error in (function (countData, colData, design, tidy = FALSE, ignoreRank = FALSE,  : \n",
      "  rownames of the colData:\n",
      "   disease3_scan | 007 | Geometric Segment,disease3_scan | 008 | Geometric Segment,disease3_scan | 009 | Geometric Segment,disease3_scan | 010 | Geometric Segment,disease3_scan | 011 | Geometric Segment,disease3_scan | 012 | Geometric Segment,disease3_scan | 013 | Geometric Segment,disease3_scan | 014 | Geometric Segment,disease3_scan | 015 | Geometric Segment,disease3_scan | 016 | Geometric Segment,disease3_scan | 017 | Geometric Segment,disease3_scan | 018 | Geometric Segment,disease3_scan | 019 | Geometric Segment,disease3_scan | 020 | Geometric Segment,disease3_scan | 021 | Geometric Segment,disease3_scan | 022 | Geometric Segment,disease3_scan | 023 | Geometric Segment,disease3_scan | 024 | Geometric Segment,disease3_scan | 025 | Geometric Segment,disease3_scan | 026 | Geometric Segment,disease3_scan | 027 | Geometric Segment,disease3_scan | 028 | Geometric Segment,disease3_scan | 029 | Geometric Segment,disease3_scan | 030 | Geometric Segment,di\n",
      "\n"
     ]
    },
    {
     "ename": "RRuntimeError",
     "evalue": "Error in (function (countData, colData, design, tidy = FALSE, ignoreRank = FALSE,  : \n  rownames of the colData:\n   disease3_scan | 007 | Geometric Segment,disease3_scan | 008 | Geometric Segment,disease3_scan | 009 | Geometric Segment,disease3_scan | 010 | Geometric Segment,disease3_scan | 011 | Geometric Segment,disease3_scan | 012 | Geometric Segment,disease3_scan | 013 | Geometric Segment,disease3_scan | 014 | Geometric Segment,disease3_scan | 015 | Geometric Segment,disease3_scan | 016 | Geometric Segment,disease3_scan | 017 | Geometric Segment,disease3_scan | 018 | Geometric Segment,disease3_scan | 019 | Geometric Segment,disease3_scan | 020 | Geometric Segment,disease3_scan | 021 | Geometric Segment,disease3_scan | 022 | Geometric Segment,disease3_scan | 023 | Geometric Segment,disease3_scan | 024 | Geometric Segment,disease3_scan | 025 | Geometric Segment,disease3_scan | 026 | Geometric Segment,disease3_scan | 027 | Geometric Segment,disease3_scan | 028 | Geometric Segment,disease3_scan | 029 | Geometric Segment,disease3_scan | 030 | Geometric Segment,di\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRRuntimeError\u001b[0m                             Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-122-83861fa8b2b6>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      4\u001b[0m                \u001b[0mdesign_matrix\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msample_df\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m                \u001b[0mdesign_formula\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'~ disease_status'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m                gene_column = 'TargetName') # <- telling DESeq2 this should be the gene ID column\n\u001b[0m\u001b[1;32m      7\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0mdds\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun_deseq\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/diffexpr/lib/python3.6/site-packages/diffexp-0.1-py3.6.egg/diffexpr/py_deseq.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, count_matrix, design_matrix, design_formula, gene_column)\u001b[0m\n\u001b[1;32m     64\u001b[0m         self.dds = deseq.DESeqDataSetFromMatrix(countData=self.count_matrix, \n\u001b[1;32m     65\u001b[0m                                         \u001b[0mcolData\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdesign_matrix\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 66\u001b[0;31m                                         design=self.design_formula)\n\u001b[0m\u001b[1;32m     67\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     68\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/diffexpr/lib/python3.6/site-packages/rpy2/robjects/functions.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    197\u001b[0m                 \u001b[0mkwargs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mr_k\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mv\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    198\u001b[0m         return (super(SignatureTranslatedFunction, self)\n\u001b[0;32m--> 199\u001b[0;31m                 .__call__(*args, **kwargs))\n\u001b[0m\u001b[1;32m    200\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    201\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/diffexpr/lib/python3.6/site-packages/rpy2/robjects/functions.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    123\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    124\u001b[0m                 \u001b[0mnew_kwargs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mconversion\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpy2rpy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 125\u001b[0;31m         \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mFunction\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__call__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mnew_args\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mnew_kwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    126\u001b[0m         \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mconversion\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrpy2py\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mres\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    127\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mres\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/diffexpr/lib/python3.6/site-packages/rpy2/rinterface_lib/conversion.py\u001b[0m in \u001b[0;36m_\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     43\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0m_cdata_res_to_rinterface\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfunction\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     44\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 45\u001b[0;31m         \u001b[0mcdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunction\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     46\u001b[0m         \u001b[0;31m# TODO: test cdata is of the expected CType\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     47\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0m_cdata_to_rinterface\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/diffexpr/lib/python3.6/site-packages/rpy2/rinterface.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    678\u001b[0m             )\n\u001b[1;32m    679\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0merror_occured\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 680\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0membedded\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mRRuntimeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_rinterface\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_geterrmessage\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    681\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mres\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    682\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mRRuntimeError\u001b[0m: Error in (function (countData, colData, design, tidy = FALSE, ignoreRank = FALSE,  : \n  rownames of the colData:\n   disease3_scan | 007 | Geometric Segment,disease3_scan | 008 | Geometric Segment,disease3_scan | 009 | Geometric Segment,disease3_scan | 010 | Geometric Segment,disease3_scan | 011 | Geometric Segment,disease3_scan | 012 | Geometric Segment,disease3_scan | 013 | Geometric Segment,disease3_scan | 014 | Geometric Segment,disease3_scan | 015 | Geometric Segment,disease3_scan | 016 | Geometric Segment,disease3_scan | 017 | Geometric Segment,disease3_scan | 018 | Geometric Segment,disease3_scan | 019 | Geometric Segment,disease3_scan | 020 | Geometric Segment,disease3_scan | 021 | Geometric Segment,disease3_scan | 022 | Geometric Segment,disease3_scan | 023 | Geometric Segment,disease3_scan | 024 | Geometric Segment,disease3_scan | 025 | Geometric Segment,disease3_scan | 026 | Geometric Segment,disease3_scan | 027 | Geometric Segment,disease3_scan | 028 | Geometric Segment,disease3_scan | 029 | Geometric Segment,disease3_scan | 030 | Geometric Segment,di\n"
     ]
    }
   ],
   "source": [
    "from diffexpr.py_deseq import py_DESeq2\n",
    "\n",
    "dds = py_DESeq2(count_matrix = df,\n",
    "               design_matrix = sample_df,\n",
    "               design_formula = '~ disease_status',\n",
    "               gene_column = 'TargetName') # <- telling DESeq2 this should be the gene ID column\n",
    "    \n",
    "dds.run_deseq() \n",
    "dds.get_deseq_result(contrast = ['SlideName','disease3','normal3'])\n",
    "res = dds.deseq_result \n",
    "res.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dds.normalized_count() #DESeq2 normalized count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dds.comparison # show coefficients for GLM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from the last cell, we see the arrangement of coefficients, \n",
    "# so that we can now use \"coef\" for lfcShrink\n",
    "# the comparison we want to focus on is 'sample_B_vs_A', so coef = 4 will be used\n",
    "lfc_res = dds.lfcShrink(coef=4, method='apeglm')\n",
    "lfc_res.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
